File: frontend/src/utils/userTheme.ts
const THEME_CLASS_PREFIX = 'theme-'

export const VALID_THEMES = new Set([
  'default',
  'light',
  'dark',
  'snowy',
  'autumn',
  'spring',
  'summer',
  'holiday',
  'ocean',
  'forest',
  'sunset',
  'midnight',
  'desert',
  'lavender',
  'rose',
  'cherry',
  'aurora',
  'storm',
  'wine',
  'mint',
])

export function normalizeTheme(theme: string | null | undefined): string {
  const value = (theme || 'default').trim().toLowerCase()
  return VALID_THEMES.has(value) ? value : 'default'
}

/** Apply site theme to the document (matches legacy ``body.theme-*`` classes). */
export function applyUserTheme(theme: string | null | undefined): string {
  const normalized = normalizeTheme(theme)
  const { body, documentElement } = document

  for (const className of [...body.classList]) {
    if (className.startsWith(THEME_CLASS_PREFIX)) {
      body.classList.remove(className)
    }
  }

  body.classList.add(`${THEME_CLASS_PREFIX}${normalized}`)
  documentElement.removeAttribute('data-theme')
  documentElement.setAttribute('data-spa-theme', normalized)
  return normalized
}

File: utils/user_theme.py
"""Effective UI theme for authenticated users (site override or preference)."""

from __future__ import annotations

THEME_ORDER = (
    "default",
    "light",
    "snowy",
    "autumn",
    "spring",
    "summer",
    "holiday",
    "ocean",
    "forest",
    "sunset",
    "midnight",
    "desert",
    "lavender",
    "rose",
    "cherry",
    "aurora",
    "storm",
    "wine",
    "mint",
    "coral",
    "sapphire",
    "honey",
    "slate",
)

THEME_CHOICES = frozenset(THEME_ORDER)

# Retired themes fall back to Default so stored preferences cannot keep them on.
REMOVED_THEMES = frozenset({"dark"})


def normalize_theme_name(value) -> str | None:
    """Map a stored theme to a current choice. Retired themes become Default."""
    if value is None:
        return None
    normalized = str(value).strip().lower()
    if not normalized:
        return None
    if normalized in REMOVED_THEMES:
        return "default"
    if normalized in THEME_CHOICES:
        return normalized
    return None


def get_effective_theme(user) -> str:
    """Return the theme class suffix (e.g. ``ocean``) for the signed-in user."""
    try:
        from models import SystemConfig

        site_override = normalize_theme_name(SystemConfig.get_value("site_theme_override"))
        if site_override:
            return site_override
    except Exception:
        pass

    pref = getattr(user, "theme_preference", None) if user is not None else None
    return normalize_theme_name(pref) or "default"


def get_site_theme_override() -> str | None:
    try:
        from models import SystemConfig

        raw = SystemConfig.get_value("site_theme_override")
        if not raw:
            return None
        return normalize_theme_name(raw)
    except Exception:
        return None
